Pre-encode headlines, captions, images using tokenizer

ArjunKrishna3367 opened the pull request:
Closes https://github.com/mahalrs/newsgen/issues/16

Uses tokenizer to encode headlines, captions, and images and saves them to new .json files (headlines_encoded.json and captions_encoded.json)

mahalrs commented on src/encode_data.py:
```diff
@@ -0,0 +1,81 @@
+    args = parser.parse_args()
+    dataset_dir = args.dataset
+
+    torch.set_grad_enabled(False)
```
No need for turning off grads here; tokenizer will handle it

mahalrs commented on src/encode_data.py:
```diff
@@ -0,0 +1,81 @@
+
+def main():
+    args = parser.parse_args()
+    dataset_dir = args.dataset
```
assert if dataset path exists and if it is a directory

mahalrs commented on src/encode_data.py:
```diff
@@ -0,0 +1,81 @@
+                    help='Directory containing VisualNews dataset')
+
+
+device = 'cuda' if torch.cuda.is_available() else 'cpu'
```
move to main()

mahalrs commented on src/encode_data.py:
```diff
@@ -0,0 +1,81 @@
+
+
+device = 'cuda' if torch.cuda.is_available() else 'cpu'
+tok = NewsgenTokenizer('../src/pretrained/exp1.ckpt', device)
```
move to main() and pass as argument to encode()

mahalrs commented on src/encode_data.py:
```diff
@@ -0,0 +1,81 @@
+                    image_batch.append(img)
+                
+                if len(headline_batch) > 0:
+                    headline_encoding = tok.encode_text_batch(headline_batch)['input_ids']
```
Store everything returned, not just 'input_ids'. Attention mask is also needed. Just append everything returned by tokenizer.

mahalrs commented on src/encode_data.py:
```diff
@@ -0,0 +1,81 @@
+                
+                if len(headline_batch) > 0:
+                    headline_encoding = tok.encode_text_batch(headline_batch)['input_ids']
+                caption_encoding = tok.encode_text_batch(caption_batch)['input_ids']
```
Store everything returned, not just 'input_ids'. Attention mask is also needed. Just append everything returned by tokenizer.

mahalrs commented:
Related to #20

mahalrs reviewed (changes requested):
Let me know when you make requested changes

ArjunKrishna3367 commented:
Fixed most of the requests. Turns out you can't index the batch encoding from the tokenizer directly, so I just pulled out the input_ids and attention_masks and added them as separate fields to the json for each article. Hopefully that works.

mahalrs commented on src/encode_data.py:
```diff
@@ -0,0 +1,86 @@
+                    image_batch.append(img)
+                
+                if len(headline_batch) > 0:
+                    headline_encoding = tok.encode_text_batch(headline_batch)
```
Are you sure .cpu() is not needed?

`headline_encoding = tok.encode_text_batch(headline_batch).cpu()`

ArjunKrishna3367 replied:
This is causing an AttributeError. They're tensors, so I can do `tok.encode_text_batch(headline_batch).to("cpu") `instead, seems to be working.

mahalrs replied:
.to('cpu') sounds good

mahalrs commented on src/encode_data.py:
```diff
@@ -0,0 +1,86 @@
+                    headline_encoding = tok.encode_text_batch(headline_batch)
+                    headline_tokens = headline_encoding['input_ids']
+                    headline_attention = headline_encoding['attention_mask']
+                caption_encoding = tok.encode_text_batch(caption_batch)
```
.cpu()

`headline_encoding = tok.encode_text_batch(caption_batch).cpu()`

mahalrs commented on src/encode_data.py:
```diff
@@ -0,0 +1,86 @@
+                caption_encoding = tok.encode_text_batch(caption_batch)
+                caption_tokens = caption_encoding['input_ids']
+                caption_attention = caption_encoding['attention_mask']
+                image_encoding = tok.encode_image_batch(image_batch)
```
.cpu()

`image_encoding = tok.encode_image_batch(image_batch).cpu()`

mahalrs reviewed (changes requested):
Minor changes

mahalrs commented on src/encode_data.py:
```diff
@@ -0,0 +1,86 @@
+    assert os.path.isdir(dataset_dir), f'{dataset_dir} is not a directory.'
+    
+    device = 'cuda' if torch.cuda.is_available() else 'cpu'
+    tok = NewsgenTokenizer('../src/pretrained/exp1.ckpt', device)
```
Better to take checkpoint path as command line argument.

mahalrs reviewed (approved):
/lgtm
